google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntervalTest.java

```diff
@@ -0,0 +1,284 @@
+  }
+
+  @Test
+  public void testParseFromString() {
+    Interval interval = Interval.parseFromString("P1Y2M3DT4H5M6.789000123S");
+    assertEquals(14, interval.months());
+    assertEquals(3, interval.days());
+    assertEquals(
+        4 * Interval.MICROS_PER_HOUR
+            + 5 * Interval.MICROS_PER_MINUTE
+            + 6 * Interval.MICROS_PER_SECOND
+            + 789000,
+        interval.micros());
+    assertEquals(123, interval.nanoFractions());
+  }
+
+  @Test
+  public void testParseFromString_InvalidString() {
+    assertThrows(SpannerException.class, () -> Interval.parseFromString("invalid"));
+    assertThrows(SpannerException.class, () -> Interval.parseFromString("P"));
+    assertThrows(SpannerException.class, () -> Interval.parseFromString("P0"));
+    assertThrows(SpannerException.class, () -> Interval.parseFromString("P0"));
+    assertThrows(SpannerException.class, () -> Interval.parseFromString("PT0.12345678912S"));
+    assertThrows(SpannerException.class, () -> Interval.parseFromString("PT.12345678912S"));
+    assertThrows(SpannerException.class, () -> Interval.parseFromString("PT1S1S"));
+    assertThrows(SpannerException.class, () -> Interval.parseFromString("P1Y2M3M"));
+  }
+
+  @Test
+  public void testToISO8601() {
+    Interval interval =
+        Interval.fromMonthsDaysMicros(
+            14,
+            3,
+            4 * Interval.MICROS_PER_HOUR
+                + 5 * Interval.MICROS_PER_MINUTE
+                + 6 * Interval.MICROS_PER_SECOND
+                + 789000);
+    assertEquals("P1Y2M3DT4H5M6.789S", interval.ToISO8601());
+  }
+
+  @Test
+  public void testToISO8601_ZeroInterval() {
+    Interval interval = Interval.zeroInterval();
+    assertEquals("P0Y", interval.ToISO8601());
+  }
+
+  @Test
+  public void testToISO8601WithNanos() {
+    Interval interval =
+        Interval.fromMonthsDaysNanos(
+            14,
+            3,
+            BigInteger.valueOf(4)
+                .multiply(Interval.NANOS_PER_HOUR)
+                .add(BigInteger.valueOf(5).multiply(Interval.NANOS_PER_MINUTE))
+                .add(BigInteger.valueOf(6).multiply(Interval.NANOS_PER_SECOND))
+                .add(BigInteger.valueOf(789123456)));
+    assertEquals("P1Y2M3DT4H5M6.789123456S", interval.ToISO8601());
+  }
+
+  @Test
+  public void testToISO8601WithOnlyNanos() {
+    Interval interval = Interval.ofNanos(BigInteger.valueOf(123456789));
+    assertEquals("PT0.123456789S", interval.ToISO8601());
+  }
+
+  @Test
+  public void testEquals() {
+    Interval interval1 = Interval.fromMonthsDaysMicros(10, 20, 30);
+    Interval interval2 = Interval.fromMonthsDaysMicros(10, 20, 30);
+    assertEquals(interval1, interval2);
+  }
+
+  @Test
+  public void testHashCode() {
+    Interval interval1 = Interval.fromMonthsDaysMicros(10, 20, 30);
+    Interval interval2 = Interval.fromMonthsDaysMicros(10, 20, 30);
+    assertEquals(interval1.hashCode(), interval2.hashCode());
+  }
```

Comment:
These should cover more variations, and also include verifications for non-equality

Reply:
Updated tests for Equality and HashCode.